browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -215,26 +223,55 @@ class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
   }
 };
 
+// Called after checking a request without any CNAME uncloaking.
+void OnShouldBlockRequestResult(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    const ResponseCallback& next_callback,
+    std::shared_ptr<BraveRequestInfo> ctx) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  if (ctx->blocked_by == kAdBlocked) {
+    brave_shields::DispatchBlockedEvent(
+        ctx->request_url, ctx->render_frame_id, ctx->render_process_id,
+        ctx->frame_tree_node_id, brave_shields::kAds);
+    next_callback.Run();
+  } else {
+    // If not blocked, the request still needs to be CNAME uncloaked.
+
+    // DoH or standard DNS queries won't be routed through Tor, so we need to
+    // skip it.
+    if (ctx->browser_context->IsTor() || ctx->did_match_important) {
+      next_callback.Run();
+    } else {
+      new AdblockCnameResolveHostClient(std::move(next_callback), task_runner,
+                                        ctx);
+    }
+  }
+}
+
+void ShouldBlockRequest(scoped_refptr<base::SequencedTaskRunner> task_runner,
+                        const ResponseCallback& next_callback,
+                        std::shared_ptr<BraveRequestInfo> ctx) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  task_runner->PostTaskAndReply(
+      FROM_HERE,
+      base::BindOnce(&ShouldBlockRequestOnTaskRunner, ctx, base::nullopt),
+      base::BindOnce(&OnShouldBlockRequestResult, task_runner, next_callback,
+                     ctx));
+}
+
 void OnBeforeURLRequestAdBlockTP(const ResponseCallback& next_callback,
                                  std::shared_ptr<BraveRequestInfo> ctx) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
   DCHECK_NE(ctx->request_identifier, 0UL);
   DCHECK(!ctx->request_url.is_empty());
   DCHECK(!ctx->initiator_url.is_empty());
 
+  DCHECK(g_brave_browser_process);
   scoped_refptr<base::SequencedTaskRunner> task_runner =
       g_brave_browser_process->ad_block_service()->GetTaskRunner();
 
   DCHECK(ctx->browser_context);
-  // DoH or standard DNS quries won't be routed through Tor, so we need to skip
-  // it.
-  if (ctx->browser_context->IsTor()) {
-    ShouldBlockAdWithOptionalCname(task_runner, std::move(next_callback), ctx,
-                                   base::nullopt);
-  } else {
-    new AdblockCnameResolveHostClient(std::move(next_callback), task_runner,
-                                      ctx);
-  }
+  ShouldBlockRequest(task_runner, std::move(next_callback), ctx);
```

Comment:
note: you may want to change the function signature to take the callback by value since you are moving it.

Reply:
I'll wait for @iefremov's comment on this one - I see `const ResponseCallback& next_callback` used across all the network delegate helper methods.

Reply:
let's keep it consistent for now. It appears to me that next_callback should become `OnceCallback` instead of the repeating one and moved everywher.